File: problem1.go

// Time Complexity : o(N)
// Space Complexity : O(H) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

//Recursive Approach
//TC O(N)
//SC O(H)
var prev *TreeNode
func isValidBST(root *TreeNode) bool {
    if root == nil{
        return false
    } 
    
    prev = nil
    return helper(root)
}

func helper(root *TreeNode) bool {

    if root == nil{
        return true
    }
    
    case1 := helper(root.Left) 
    
    if prev != nil && prev.Val >= root.Val{
        return false
    }
    
    prev = root
    
    case2 := helper(root.Right)
    
    return  case1 && case2 
}



///////////////////////////////////////////////////////////////////////////////
//Iterative Approach
//TC O(N)
//SC O(H)
var prev *TreeNode
func isValidBST(root *TreeNode) bool {
    if root == nil{
        return false
    } 
    
    stack := []*TreeNode{}
    prev = nil
    for root != nil || len(stack) > 0{
        for root != nil{
            stack = append(stack,root)
            root = root.Left
            
        }
        root = stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        if prev != nil && prev.Val >= root.Val{
                return false
            }
        prev = root
        root = root.Right
    }
    
    return true
}


///////////////////////////////////////////////////////////////////////////////////////
//By min max variables
//TC - O(N)
//SC  - O(H)

func isValidBST(root *TreeNode) bool {

    var min *int = nil
    var max *int  = nil
    
    return helper(root,min,max)
}


func helper(root *TreeNode,min,max *int) bool{
    if root == nil{
        return true
    }
    
    if max != nil && root.Val >= *max{
        return false
    }
    
    if min != nil && root.Val <= *min{
        return false
    }
    
    return helper(root.Left,min,&root.Val) && helper(root.Right,&root.Val,max)
}

File: Problem_1.js
// https://leetcode.com/problems/validate-binary-search-tree/

// Time Complexity : O(n) n is the tree nodes
// Space Complexity : O(h) h is the tree height
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : When returning true/false in recursion


// Your code here along with comments explaining your approach
// dfsRange
// Maintain a range of numbers between which the root's value shold lie

// dfsRecursive
// Inorder traversal of a BST will give numbers in increasing order.
// We keep a track of the previous number and if it is more or equal to current, the BST is invalid

var dfsRange = (root, low, high) => {
    // Base
    if (root === null)
        return true;
    // Logic
    if (low !== null && root.val <= low) return false;
    if (high !== null && root.val >= high) return false;

    return dfs(root.left, low, root.val) && dfs(root.right, root.val, high);
}

// let prev;
// let isValid;
// var recursiveDfs = (root) => {
//     if (root === null)
//         return true;

//     if (!dfs(root.left))
//         return false;
//     if (prev != null && prev >= root.val) {
//         isValid = false;
//     }
//     prev = root.val;
//     return dfs(root.right);
// }

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
    if (root === null) {
        return true;
    }
    // Range
    return dfsRange(root, null, null);
    // Resursive
    // prev = null;
    // isValid = true
    // recursiveDfs(root);
    // return isValid;
};